bin/varnishd/cache/cache_expire.c

```diff
@@ -105,25 +105,45 @@ exp_mail_it(struct objcore *oc, uint8_t cmds)
 {
 	CHECK_OBJ_NOTNULL(oc, OBJCORE_MAGIC);
 	assert(oc->refcnt > 0);
+	AZ(cmds & OC_EF_REFD);
 
-	Lck_Lock(&exphdl->mtx);
-	if ((cmds | oc->exp_flags) & OC_EF_REFD) {
+	Lck_AssertHeld(&exphdl->mtx);
+
+	if (oc->exp_flags & OC_EF_REFD) {
```

Comment:
It seems this is already guaranteed except for the call from `EXP_Rearm()`.

Would it make sense to turn this into `AN(oc->exp_flags & OC_EF_REFD)` and have `EXP_Rearm()` only call `exp_mail_it()` if OC_EF_REFD is set?

Note: I understand that we need to operate on `OC_EF_NEW` under the lock, but I do not see how we could race for `OC_EF_REFD` ?

Reply:
It is also a concern from EXP_Remove. OC_EF_REFD is first tested without the lock in both EXP_Remove and EXP_Rearm., because it can not be gained once lost, and early return is done if the flag is missing. That way the mutex is not needed for objects not on expiry, e.g. transient and passes. But if it was found before getting the mutex, it needs to be tested again after getting the mutex. exp_mail_it handles that for both EXP_Rearm and EXP_Remove.

Reply:
Thank you, I see it now.